File: 代码随想录/08_回溯算法/04_组合总和 III/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> res;
    vector<int> path;
    int n, k;
    vector<vector<int>> combinationSum3(int k, int n) {
        this->k = k;
        this->n = n;
        dfs(path, 0, 1); 
        return res;
    }

    void dfs(vector<int>& path, int sum, int u) {
        if (sum > n) return; // 剪枝
        if (sum == n && path.size() == k) {
            res.push_back(path);
            return;
        }
        for (int i = u; i <= 9; i++) {
            path.push_back(i);
            sum += i;
            dfs(path, sum, i + 1);
            path.pop_back();
            sum -= i;
        }
    }
};

File: 代码随想录/08_回溯算法/04_组合总和 III/Solution.java
import java.util.ArrayList;
import java.util.List;

// https://leetcode.cn/problems/combination-sum-iii/
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    int k, n;

    public List<List<Integer>> combinationSum3(int k, int n) {
        this.k = k;
        this.n = n;
        dfs(new ArrayList<>(), 0, 1); 
        return res;
    }

    void dfs(List<Integer> path, int sum, int u) {
        if (sum > n) return; // 剪枝
        if (sum == n && path.size() == k) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = u; i <= 9; i++) {
            sum += i;
            path.add(i);
            dfs(path, sum, i + 1); 
            sum -= i;
            path.remove(path.size() - 1);
        }
    }
}